src/storage/txn/commands/mod.rs

```diff
@@ -377,6 +384,36 @@ impl Command {
         }
     }
 
+    pub fn read_command_mut<E: Engine>(&mut self) -> &mut dyn ReadCommand<E> {
```

Comment:
If the performance regression caused by copy is really serious, we can make `ReadCommand::process_read` take `self` instead of `&mut self` and do static dispatch ie.`t.process_read` manually here.

Reply:
Where is the copy? I see that we convert to a trait object, but it seems to only be a reference copy?

Reply:
Looking at this code, it seems to me that we Command should just be a trait, rather than an enum, then we could skip these methods and the as_ref/as_mut methods too. Maybe we could get rid of TypedCommand too? If this is a lot of work, lets leave it for another PR.

Reply:
> Where is the copy? I see that we convert to a trait object, but it seems to only be a reference copy?

The copies are in the implementations, in the old approach, we "consume" the command and all its fields, but now what we have is a reference to the command, so I have to copy some of a commands' fields.

Reply:
> Looking at this code, it seems to me that we Command should just be a trait, rather than an enum, then we could skip these methods and the as_ref/as_mut methods too. Maybe we could get rid of TypedCommand too? If this is a lot of work, lets leave it for another PR.

That's also my final target, but it may affect some code about the `Task` struct, the scheduler and many other things.
I'll open another pr for that.

Reply:
I see, so then why can't we use `self` rather than `&mut self`?

Reply:
> I see, so then why can't we use `self` rather than `&mut self`?

Because we do need to do dynamic dispatch. We don’t know the actual type of `self` when call the read/write process unless we use the Command enum, which we are planning to remove.

Reply:
I would leave it as just `self` for now and we can deal with the dispatch issue later. We could use `Box` rather than `&` since I'm sure we can't leave the commands on the stack in any case.